README.md
# rfrx

[![Test](https://github.com/nim65s/rfrx/actions/workflows/test.yml/badge.svg)](https://github.com/nim65s/rfrx/actions/workflows/test.yml)
[![Release](https://github.com/nim65s/rfrx/actions/workflows/release.yml/badge.svg)](https://github.com/nim65s/rfrx/actions/workflows/release.yml)
[![PyPI version](https://badge.fury.io/py/rfrx.svg)](https://pypi.org/project/rfrx)
[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/nim65s/rfrx/main.svg)](https://results.pre-commit.ci/latest/github/nim65s/rfrx/main)
[![codecov](https://codecov.io/gh/nim65s/rfrx/branch/main/graph/badge.svg?token=BLGISGCYKG)](https://codecov.io/gh/nim65s/rfrx)

[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/charliermarsh/ruff/main/assets/badge/v1.json)](https://github.com/charliermarsh/ruff)

Receive RF frames from some remote controllers.

Available Decoder / Readers:
- SBUS
- Pro-Tronik PTR-6A v2

:warning: Don't forget to invert its electrical signal, eg. with a NAND gate ! :warning:

## Configuration

Could be done in python calls, or by setting the following environment variables:

### SBUS

- `RFRX_PORT`: default to `/dev/ttyS0`
- `RFRX_TIMEOUT`: default to 1
- `RFRX_N_CHANS`: default to 16, number of useful channels for your case
- `RFRX_RUNNING`: default to `ON`, to keep reading and procesing data
- `RFRX_RETRY`: default to `ON`, to close and reopen the port again when something went wrong
- `RFRX_LOG_LEVEL`: default to `WARNING`

### ProTronik

Minimum, Middle and Maximum values read on your Right and Left Horizontal and Vertical joysticks

- `RFRX_RX_MIN`: default to 192
- `RFRX_RX_MID`: default to 992
- `RFRX_RX_MAX`: default to 1796
- `RFRX_RY_MIN`: default to 302
- `RFRX_RY_MID`: default to 1100
- `RFRX_RY_MAX`: default to 1900
- `RFRX_LY_MIN`: default to 180
- `RFRX_LY_MID`: default to 980
- `RFRX_LY_MAX`: default to 1779
- `RFRX_LX_MIN`: default to 192
- `RFRX_LX_MID`: default to 992
- `RFRX_LX_MAX`: default to 1790


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- setup coverage

## [v2.1.2] - 2023-05-26

- changelog: typo

## [v2.1.1] - 2023-05-26

- fix release script and changelog

## [v2.1.0] - 2023-05-26

- add unittests with a couple real frames from pro-tronik
- add urls to pyproject

## [v2.0.0] - 2023-05-26

- protronik: rename h/v → x/y
- configuration through `RFRX_` environment variables
- configure log level if no root logging handler is defined
- add release script

## [v1.1.0] - 2023-05-25

- add `SbusReader.running` parameter to stop main loop on request

## [v1.0.1] - 2023-05-25

- fix Release

## [v1.0.0] - 2023-05-25

- Initial Release

[Unreleased]: https://github.com/nim65s/rfrx/compare/v2.1.2...HEAD
[v2.1.2]: https://github.com/nim65s/rfrx/compare/v2.1.1...v2.1.2
[v2.1.1]: https://github.com/nim65s/rfrx/compare/v2.1.0...v2.1.1
[v2.1.0]: https://github.com/nim65s/rfrx/compare/v2.0.0...v2.1.0
[v2.0.0]: https://github.com/nim65s/rfrx/compare/v1.1.0...v2.0.0
[v1.1.0]: https://github.com/nim65s/rfrx/compare/v1.0.1...v1.1.0
[v1.0.1]: https://github.com/nim65s/rfrx/compare/v1.0.0...v1.0.1
[v1.0.0]: https://github.com/nim65s/rfrx/releases/tag/v1.0.0


tests/__init__.py
"""Test RFRX module."""


tests/test_sbus.py
"""Test sbus decoder."""
from unittest import TestCase

from rfrx.sbus import SbusDecoder


class TestSbus(TestCase):
    """Test sbus decoder."""

    def test_frame(self):
        """Test sbus decoder."""
        frame = (
            b'\x0f\xe0c"\xf1\xc0\x07>\x7f\x83\x0f|\x00\x00'
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
        )
        sbus = SbusDecoder(frame)
        self.assertFalse(sbus.ch17)
        self.assertFalse(sbus.ch18)
        self.assertFalse(sbus.ch18)
        self.assertEqual(sbus.chans[0], 992)

        frame = (
            b"\x0fCb\xe2\xf0\xc0\x07\x0c\x80\x83\x0f|\x00\x00"
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
        )

        sbus = SbusDecoder(frame)
        self.assertFalse(sbus.ch17)
        self.assertFalse(sbus.ch18)
        self.assertFalse(sbus.ch18)
        self.assertEqual(sbus.chans[0], 579)


tests/test_protronik.py
"""Test pro-tronik decoder."""
from unittest import TestCase

from rfrx.protronik import ProTronikDecoder


class TestProTronik(TestCase):
    """Test pro-tronik decoder."""

    def test_frame(self):
        """Test pro-tronik decoder."""
        frame = (
            b'\x0f\xe0c"\xf1\xc0\x07>\x7f\x83\x0f|\x00\x00'
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
        )
        sbus = ProTronikDecoder(frame)
        self.assertEqual(sbus.rx, 0.0)
        self.assertEqual(sbus.ch5, 1)

        frame = (
            b"\x0fCb\xe2\xf0\xc0\x07\x0c\x80\x83\x0f|\x00\x00"
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
        )

        sbus = ProTronikDecoder(frame)
        self.assertEqual(sbus.rx, -0.51625)
        self.assertEqual(sbus.ch5, 2)


rfrx/sbus.py
"""Process SBUS data."""
from logging import basicConfig, getLogger, root
from os import environ
from time import sleep

import serial

LOGGER = getLogger("rfrx.sbus")

_FALSY = ("0", "NO", "OFF", "FALSE")
PORT = environ.get("RFRX_PORT", "/dev/ttyS0")
RUNNING = environ.get("RFRX_RUNNING", "ON") not in _FALSY
RETRY = environ.get("RFRX_RETRY", "ON") not in _FALSY
TIMEOUT = int(environ.get("RFRX_TIMEOUT", 1))
N_CHANS = int(environ.get("RFRX_N_CHANS", 16))
LOG_LEVEL = environ.get("RFRX_LOG_LEVEL", "WARNING").upper()


class SbusError(Exception):
    """Exception class for SbusDecoder."""

    pass


class SbusDecoder:
    """Decode SBUS frames."""

    def __init__(self, frame, n_chans=N_CHANS):
        """Parse a frame with `n_chans` used channels."""
        LOGGER.info("parsing frame")
        self.frame = frame
        self.val = int.from_bytes(frame, byteorder="little")
        if not (0 <= n_chans <= 16):
            err = (
                "Wrong number of channels. "
                f"Expected 0 <= n_chans <= 16, got {n_chans}."
            )
            raise ValueError(err)
        self.n_chans = n_chans
        self.chans = [None] * n_chans
        self.failsafe = None
        self.frame_lost = None
        self.ch17 = None
        self.ch18 = None

        self.check()
        LOGGER.debug(f"0x{self.val:025x}")
        self.parse()
        LOGGER.debug("chans: %s" % self.chans)
        LOGGER.debug("failsafe: %s" % self.failsafe)
        LOGGER.debug("frame lost: %s" % self.frame_lost)
        LOGGER.debug("ch 18: %s" % self.ch18)
        LOGGER.debug("ch 17: %s" % self.ch17)

    def check(self):
        """Check if the frame is in correct format."""
        if len(self.frame) != 25:
            err = f"Wrong frame length. Expected 25, got {len(self.frame)}."
            raise SbusError(err)
        if self.frame[0] != 0x0F:
            err = f"Wrong header. Expected 0x0F, got {self.frame[0]}."
            raise SbusError(err)
        if self.frame[24] != 0x00:
            err = f"Wrong end byte. Expected 0x00, got {self.frame[24]}."
            raise SbusError(err)

    def parse(self):
        """Decode `n_chans` channels and 4 bits in Byte 23."""
        data = self.val >> 8  # ignore header
        for chan in range(self.n_chans):
            self.chans[chan] = data & 0b11111111111
            data >>= 11

        self.failsafe = self.frame[23] & 0x10
        self.frame_lost = self.frame[23] & 0x20
        self.ch18 = self.frame[23] & 0x40
        self.ch17 = self.frame[23] & 0x80

    def __str__(self):
        """Show decoded data."""
        data = [*self.chans, self.failsafe, self.frame_lost, self.ch18, self.ch17]
        return " ".join(f"{c:4}" for c in data)


class SbusReader:
    """Bind to a serial port, read SBUS frames and process them."""

    decoder = SbusDecoder

    def __init__(self, port=PORT, running=RUNNING, retry=RETRY, timeout=TIMEOUT):
        """Configure the serial port parameters."""
        if len(root.handlers) == 0:
            basicConfig(level=LOG_LEVEL)
        self.port = port
        self.running = running
        self.retry = retry
        self.timeout = timeout

        self.run()

    def run(self):
        """Run main loop."""
        while self.running and self.retry:
            with serial.Serial(
                port=self.port,
                baudrate=100_000,
                bytesize=serial.EIGHTBITS,
                # TODO: broken on raspbian buster & pyserial 3.5
                # works well on same setup but with pyserial 3.4…
                # parity=serial.PARITY_EVEN,
                stopbits=serial.STOPBITS_TWO,
                timeout=self.timeout,
            ) as ser:
                while self.running:
                    try:
                        data = ser.read(25)
                        if len(data) == 0:
                            err = (
                                f"no data on {self.port} "
                                f"in the last {self.timeout} seconds."
                            )
                            LOGGER.warning(err)
                        else:
                            frame = self.decoder(data)
                            self.process(frame)
                    except SbusError as e:
                        LOGGER.error(f"SBUS error: {e}")
                        break
            LOGGER.warning("sleeping a bit.")
            sleep(1)

    def process(self, frame):
        """Override this for your application."""
        print(frame, frame.chans)


rfrx/protronik.py
"""Process protronik data."""

from logging import getLogger
from os import environ

from .sbus import SbusDecoder, SbusReader

LOGGER = getLogger("rfrx.protronik")

RX_MIN, RX_MID, RX_MAX = (
    int(environ.get("RFRX_RX_MIN", 192)),
    int(environ.get("RFRX_RX_MID", 992)),
    int(environ.get("RFRX_RX_MAX", 1796)),
)
RY_MIN, RY_MID, RY_MAX = (
    int(environ.get("RFRX_RY_MIN", 302)),
    int(environ.get("RFRX_RY_MID", 1100)),
    int(environ.get("RFRX_RY_MAX", 1900)),
)
LY_MIN, LY_MID, LY_MAX = (
    int(environ.get("RFRX_LY_MIN", 180)),
    int(environ.get("RFRX_LY_MID", 980)),
    int(environ.get("RFRX_LY_MAX", 1779)),
)
LX_MIN, LX_MID, LX_MAX = (
    int(environ.get("RFRX_LX_MIN", 192)),
    int(environ.get("RFRX_LX_MID", 992)),
    int(environ.get("RFRX_LX_MAX", 1790)),
)


class ProTronikDecoder(SbusDecoder):
    """Decode S-BUS frames from Pro-Tronik PTR-6A v2."""

    def __init__(self, frame):
        """Decode SBUS, then parse it according to Pro-Tronik use."""
        super().__init__(frame, n_chans=6)

        # calibration: minimum, middle, maximum, name
        rx = RX_MIN, RX_MID, RX_MAX, "right horizontal"
        ry = RY_MIN, RY_MID, RY_MAX, "right vertical"
        ly = LY_MIN, LY_MID, LY_MAX, "left vertical"
        lx = LX_MIN, LX_MID, LX_MAX, "left horizontal"

        def scale(val, mini, middle, maxi, name):
            """Convert read value to -1.0 -> 1.0."""
            if val < mini:
                LOGGER.debug(f"{name} value lower than min: {val} < {mini}")
                return -1.0
            if val > maxi:
                LOGGER.debug(f"{name} value higher than max: {val} > {maxi}")
                return 1.0
            if val < middle:
                return (val - mini) / (middle - mini) - 1.0
            return (val - middle) / (maxi - middle)

        self.rx = +scale(self.chans[0], *rx)
        self.ry = -scale(self.chans[1], *ry)
        self.ly = -scale(self.chans[2], *ly)
        self.lx = +scale(self.chans[3], *lx)

        self.ch5 = 2 if self.chans[4] < 500 else 1 if self.chans[4] < 1500 else 0
        self.ch6 = 0 if self.chans[5] < 1000 else 1

    def __str__(self):
        """Show decoded data."""
        return (
            f"{self.rx:+.3f} {self.ry:+.3f} {self.ly:+.3f} {self.lx:+.3f} "
            f"{self.ch5} {self.ch6}"
        )


class ProTronikReader(SbusReader):
    """Bind to a serial port, read SBUS pro-tronik frames and process them."""

    decoder = ProTronikDecoder
